graphql/src/main/java/com/linecorp/armeria/internal/server/graphql/GraphqlDocServicePlugin.java

```diff
@@ -0,0 +1,151 @@
+import com.linecorp.armeria.server.docs.ServiceSpecification;
+import com.linecorp.armeria.server.docs.TypeSignature;
+import com.linecorp.armeria.server.graphql.protocol.AbstractGraphqlService;
+
+/**
+ * A {@link DocServicePlugin} implementation that supports the {@link AbstractGraphqlService}.
+ */
+public final class GraphqlDocServicePlugin implements DocServicePlugin {
+
+    @VisibleForTesting
+    static final TypeSignature STRING = TypeSignature.ofBase("string");
+    @VisibleForTesting
+    static final TypeSignature MAP = TypeSignature.ofBase("map");
+    @VisibleForTesting
+    static final TypeSignature JSON = TypeSignature.ofBase("json");
+
+    static final String DEFAULT_METHOD_NAME = "doPost";
+
+    @Override
+    public String name() {
+        return "graphql";
+    }
+
+    @Override
+    public Set<Class<? extends Service<?, ?>>> supportedServiceTypes() {
+        return ImmutableSet.of(AbstractGraphqlService.class);
+    }
+
+    @Override
+    public ServiceSpecification generateSpecification(Set<ServiceConfig> serviceConfigs,
+                                                      DocServiceFilter filter) {
+        requireNonNull(serviceConfigs, "serviceConfigs");
+        requireNonNull(filter, "filter");
+
+        final Map<Class<?>, Set<MethodInfo>> methodInfos = new HashMap<>();
+        final Map<Class<?>, String> serviceDescription = new HashMap<>();
+        serviceConfigs.forEach(sc -> {
+            final AbstractGraphqlService service = sc.service().as(AbstractGraphqlService.class);
+            if (service != null) {
+                final String className = service.getClass().getName();
+                final String methodName = DEFAULT_METHOD_NAME;
+                if (!filter.test(name(), className, methodName)) {
+                    return;
+                }
+                addMethodInfo(methodInfos, sc.virtualHost().hostnamePattern(), service, sc.route());
+            }
+        });
+
+        return generate(serviceDescription, methodInfos);
+    }
+
+    private static void addMethodInfo(Map<Class<?>, Set<MethodInfo>> methodInfos,
+                                      String hostnamePattern, AbstractGraphqlService service, Route route) {
+        final EndpointInfo endpoint = endpointInfo(route, hostnamePattern);
+        final String name = DEFAULT_METHOD_NAME;
+        final List<FieldInfo> fieldInfos = fieldInfos();
+        final Class<?> clazz = service.getClass();
+        final MethodInfo methodInfo = new MethodInfo(
+                name, JSON, fieldInfos, ImmutableList.of(), // Ignore exceptions.
+                ImmutableList.of(endpoint), HttpMethod.POST, null);
+        methodInfos.computeIfAbsent(clazz, unused -> new HashSet<>()).add(methodInfo);
+    }
+
+    private static EndpointInfo endpointInfo(Route route, String hostnamePattern) {
+        final List<String> paths = route.paths();
+        return EndpointInfo.builder(hostnamePattern, RouteUtil.EXACT + paths.get(0))
```

Comment:
note: prefixing paths with `exact:` breaks `Copy as a curl command` if multiple services are defined (this is irrelevant of this PR, and occurred for annotated services as well). Will create a separate issue for this.

https://github.com/line/armeria/blob/c0b49a8803e3ca815a65fd5d37c76f33de5b7cdd/docs-client/src/lib/transports/transport.ts#L147